docs(transforms): Added failed route event routing highlight article (#10052)
vectordotdev/vector: website/content/en/highlights/2021-11-16-failed-event-routing.md

---
date: "2021-11-16"
title: "Initial support for routing failed events released"
description: "Routing failed events from the `remap` transform"
authors: ["barieom", "lukesteensen"]
pr_numbers: [9417, 9169]
release: "0.18.0"
hide_on_release_notes: false
badges:
  type: new feature
---

We've released a new feature that enables users to route failed events through separate pipelines.

Previously, when Vector encountered an event that failed a given transformation, the event was either dropped or was forwarded to the next step of the process. 

> With this new release, Vector enables events that fail to process down a different pipeline without any manual work-around to catch errors, tag, fanout, or filter. In other words, users can now handle failed events in a way the user sees fit, such as routing the failed events to another sink for storage, inspection, and replay. For clarification, we've only added support to the `remap` transform thus far, and unit tests do not currently support assertions on failed events. That said, more support for failed events is actively under development so keep your eyes peeled for announcements down the line. 

Reviewer: ```suggestion
With this new release, Vector enables events that fail to process down a different pipeline without any manual work-around to catch errors, tag, fanout, or filter. In other words, users can now handle failed events in a way the user sees fit, such as routing the failed events to another sink for storage, inspection, and replay. We are piloting this feature with the `remap` transform but plan to roll this out to other transforms in the future so keep your eyes peeled for announcements down the line. Note: unit tests do not currently support assertions on failed events but this is also in the works.
```

> As an example, the `dropped` output can be used if you want to capture events that failed during processing and re-direct the failed events to a separate sink from the data that was processed successfully. In a given config below:

Reviewer: ```suggestion
This feature is best illustrated by example. In the following example you can see how you can now use a new  `dropped` output from the `remap` transform to capture and route events that fail to be processed to a separate sink from the data that was processed successfully.

Let's start with this configuration:
```

``` toml
[sources.in]
  type = "generator"
  format = "shuffle"
  interval = 1.0
  lines = [
    '{ "message": "valid message", "foo": "bar"}',
    '{ "message": "valid message", "foo": "baz"}',
    'invalid message',
  ]

[transforms.my_remap]
  type = "remap"
  inputs = ["in"]
  drop_on_error = true
  drop_on_abort = true
  reroute_dropped = true
  source = """
    . |= object!(parse_json!(.message))
    if .foo == "baz" {
        abort
    }
    .processed = true
  """

[sinks.foo]
  type = "console"
  inputs = ["my_remap"]
  encoding.codec = "json"

[sinks.bar]
  type = "console"
  inputs = ["my_remap.dropped"]
  encoding.codec = "json"
``` 

> This would emit the following output:

Reviewer: ```suggestion
This would emit the following output:

```

> ``` 

Reviewer: ```suggestion
``` json
```

---
date: "2021-11-16"

> title: "Failed event routing feature released"

Reviewer: ```suggestion
title: "Initial support for routing failed events released"
```

Example of updating the language a bit to reflect that this is a beta feature under active development.

> description: "A guide for using the new pipelines functionality"

Reviewer: ```suggestion
description: "Routing failed events from the `remap` transform"
```

authors: ["barieom", "lukesteensen"]

> pr_numbers: []

Reviewer: ```suggestion
pr_numbers: [9417, 9169]
```

``` toml
[sources.in]
  type = "generator"
  format = "shuffle"
  interval = 1.0
  lines = [
    '{ "message": "valid message", "foo": "bar"}',
    '{ "message": "valid message", "foo": "baz"}',
    'invalid message',
  ]

[transforms.my_remap]
  type = "remap"
  inputs = ["in"]
  drop_on_error = true
  drop_on_abort = true
  reroute_dropped = true
  source = """
    . |= object!(parse_json!(.message))
    if .foo == "baz" {
        abort
    }
    .processed = true
  """

[sinks.foo]
  type = "console"
  inputs = ["my_remap"]
  encoding.codec = "json"

[sinks.bar]
  type = "console"
  inputs = ["my_remap.dropped"]
  encoding.codec = "json"
``` 

This would emit the following output:
``` 
  {
  "foo": "bar",
  "message": "valid message",
  "processed": true,
  "timestamp": "2021-11-09T16:11:47.330713806Z"
}
{
  "foo": "bar",
  "message": "valid message",
  "processed": true,
  "timestamp": "2021-11-09T16:11:48.330756592Z"
}
{
  "message": "invalid message",

>   "metadata": {

Reviewer: I think we should include a note in here that this `metadata` key is configurable via https://master.vector.dev/docs/reference/configuration/global-options/#log_schema.metadata_key

    "dropped": {
      "component_id": "my_remap",
      "component_type": "remap",
      "component_kind": "transform",
      "message": "function call error for \"object\" at (9:39): function call error for \"parse_json\" at (17:38): unable to parse json: expected value at line 1 column 1",
      "reason": "error"
    }
  },
  "timestamp": "2021-11-09T16:11:49.330157298Z"
}
{
  "message": "{ \"message\": \"valid message\", \"foo\": \"baz\"}",
  "metadata": {
    "dropped": {
      "component_id": "my_remap",
      "component_type": "remap",
      "component_kind": "transform",
      "message": "aborted",
      "reason": "abort"
    }
  },
  "timestamp": "2021-11-09T16:11:50.329966720Z"
}
```

> Events that either caused an error or were aborted are written out by the `bar` console sink with the relevant metadata added; on the other hand, events that were valid JSON were processed and output by the `foo` console sink. 

Reviewer: Maybe we could add a link below this paragraph to the docs for this feature?

https://master.vector.dev/docs/reference/configuration/transforms/remap/

> For our next steps, we'll be looking to add `filtering` functionality, which will enable events to be passed to the next processor if a condition is not. In the meantime, if you any feedback for us, let us know on our [Discord chat][] or on [Twitter][]!

Reviewer: ```suggestion
We will be continuing to expand support for routing failed events from other transforms like `filter`. In the meantime, if you any feedback for us, let us know on our [Discord chat][] or on [Twitter][]!
```
